File: Chess.Game/Pieces/Rook.cs
using System;
using System.Collections.Generic;

namespace Chess.Game.Pieces
{
    public class Rook : IPiece
    {
        public int Moves { get; set; }
        public Color Color { get; set; }

        public IList<Location> GetLegalMoves(Location currentLocation)
        {
            throw new NotImplementedException();
        }

        public bool IsMoveLegal(Location current, Location newLocation)
        {
            if (newLocation.CurrentColor == Color)
                return false;

            if (current.File != newLocation.File && current.Rank != newLocation.Rank)
                return false;

            var legal = true;

            if (current.File != newLocation.File)
            {
                if (newLocation.File > current.File)
                {
                    var location = current.Right;
                    while (location != newLocation && legal)
                    {
                        if (location.CurrentColor != null)
                            legal = false;
                        else
                            location = location.Right;
                    }
                }
                else
                {
                    var location = current.Left;
                    while (location != newLocation && legal)
                    {
                        if (location.CurrentColor != null)
                            legal = false;
                        else
                            location = location.Left;
                    }
                }
            }
            else
            {
                if (newLocation.Rank > current.Rank)
                {
                    var location = current.Forward;
                    while (location != newLocation && legal)
                    {
                        if (location.CurrentColor != null)
                            legal = false;
                        else
                            location = location.Forward;
                    }
                }
                else
                {
                    var location = current.Back;
                    while (location != newLocation && legal)
                    {
                        if (location.CurrentColor != null)
                            legal = false;
                        else
                            location = location.Back;
                    }
                }
            }

            return legal;
        }
    }
}

File: Chess.Web/ClientApp/pieces/Rook.ts
import { Location } from '../store/Game';
import Piece from './Piece';
import Color from './Color';
import { isStraightMoveLegal } from './MoveHelper';

class Rook implements Piece {
    color: Color;
    moves: number;

    constructor(moves: number, color: Color) {
        this.moves = moves;
        this.color = color;
    }

    isMoveLegal(current: Location, newLocation: Location, board: Location[][]) {
        if (newLocation.piece && newLocation.piece.color == this.color)
            return false;

        const legal = isStraightMoveLegal(current, newLocation, board);

        return legal;
    };

    getImageName() {
        const image = this.color == Color.White ? "white_rook" : "black_rook";

        return image;
    }
}

export default Rook;
